integration_tests/src/main/python/hash_aggregate_test.py

```diff
@@ -774,6 +774,15 @@ def test_arithmetic_reductions(data_gen):
                 'avg(a)'),
             conf = _no_nans_float_conf)
 
+@ignore_order(local=True)
+@pytest.mark.parametrize('data_gen',
+                         all_gen + [ArrayGen(g) for g in all_gen] + struct_gens_sample, ids=idfn)
+def test_groupby_first_last(data_gen):
+    gen_fn = [('a', RepeatSeqGen(LongGen(), length=20)), ('b', data_gen)]
+    agg_fn = lambda df: df.groupBy('a').agg(
+            f.first('b'), f.last('b'), f.first('b', True), f.last('b', True))
+    assert_gpu_and_cpu_are_equal_collect(lambda spark : agg_fn(gen_df(spark, gen_fn)))
```

Comment:
nit: first and last are not deterministic when they are run in a real distributed setup.  There are times when this test is run that way. It would probably be best to do a `.coalesce(1)` on the input data so that the input order is deterministic for those cases.

Reply:
Just for my understanding, will sort also make the results deterministic in those cases? 

@sperlingxx should we also add a test for Array[Struct]??

Reply:
The issue is that at least half of the aggregation happens after a shuffle, and the order that upstream task data shows up in a down stream task is not deterministic.

If I have data like
|a|b|
|-|-|
|1|1|
|1|2|
|1|3|
|1|4|

and two tasks, then the data likely would be divided up so that you get

Task 1
|a|b|
|-|-|
|1|1|
|1|2|

Task 2
|a|b|
|-|-|
|1|3|
|1|4|

The result of group by a first(b) from each of these tasks would be

Task 1
|a|b|
|-|-|
|1|1|

Task 2
|a|b|
|-|-|
|1|3|

Now when Task 3 starts that is going to pull in all of the data for `a == 1` it will start fetching from several upstream tasks in parallel.  If Task 1 shows up first you get data like.
|a|b|
|-|-|
|1|1|
|1|3|

and the file result for `first(b)` is 1, but if Task 2 wins and shows up first you get data like
|a|b|
|-|-|
|1|3|
|1|1|

and `first(b)` ends up being 3. 

Sorting the data after that makes no difference. 1 != 3  We get away with this most of the time on local mode tests because in the Spark shuffle there is an optimization for local reads That puts them first in the queue, and it really is a queue so the order is maintained. As soon as you go to a distributed setup you can get different answers from one run to the next, even on the CPU.  By doing a coalesce(1) we are only going to have a single task so it removes the ambiguity.